Add GDExtension function to construct StringName directly from `char*`

Bromeon opened the pull request:
So far, an indirection via String was necessary, causing at least 2 allocations and copies (String; String inside StringName). Since StringNames often refer to string literals, this allows them to be directly constructed from C strings.

It also provides the `is_static` flag: when the source has static storage duration, no copy/allocation will be needed. However, the extension developer needs to uphold this lifetime guarantee.

Some things need to be discussed:
1. Does this actually support UTF-8? I've only seen ASCII StringNames in the wild...
2. I adopted the name from the `String` constructor function, feel free to suggest a better one (especially if 1 doesn't hold).
3. Are the memory assumptions correct?
    * If a static StringName is created, its destructor should not be invoked?
    * Non-static ones are reference-counted as usual?

dsnopek commented:
Thanks! The overall idea here looks great :-)

> 1. Does this actually support UTF-8? I've only seen ASCII StringNames in the wild...

That's a really good question. Looking through the code, it seems as though `StringName`s created from `char *` may be interpreted as latin-1? It definitely doesn't seem like utf-8, though - `String::hash()` is used to find the right data in the table, and it's basically just casting the `char`s to `uint32_t` (as opposed to trying to interpret the multi-byte sequences from utf-8).

> 2. I adopted the name from the String constructor function, feel free to suggest a better one (especially if 1 doesn't hold).

Maybe `gdextension_string_name_new_from_chars()`? Or `gdextension_string_name_new_from_latin1_chars()` if it's really latin-1 and not just plain ASCII.

> If a static StringName is created, its destructor should not be invoked?

This I'm really not sure about this one. I think it may still be OK to invoke the destructor? I'm actually quite confused about how `StringName::_Data::static_count` is being used. This needs more investigation or an expert opinion.

dsnopek commented on core/extension/gdextension_interface.h:
```diff
@@ -1526,6 +1526,21 @@ typedef void (*GDExtensionInterfaceStringOperatorPlusEqWcstr)(GDExtensionStringP
+
+/**
+ * @name string_name_new_with_utf8_chars
+ * @since 4.1
```
This is probably actually Godot 4.2 material -- 4.1 is about to go into RC next week!

Bromeon committed:
GDExtension: construct StringName directly from char*

So far, an indirection via String was necessary, causing at least 2 allocations and copies (String; String inside StringName).
Since StringNames often refer to string literals, this allows them to be directly constructed from C strings.

There are two formats: Latin-1 and UTF-8.

The Latin-1 constructor also provides the `p_is_static` flag: when the source has static storage duration, no copy/allocation will be needed.
However, the extension developer needs to uphold this lifetime guarantee.

Bromeon commented:
It looks like the encoding depends on whether the `StringName` is constructed statically or not. `string_name_new_with_chars` is now named more generally to accomodate for both encodings.

I mentioned the semantics in the docs, but would appreciate a 2nd opinion:

>  If `is_static` is true, then:
>  - The StringName will reuse the `p_contents` buffer instead of copying it.
>    You must guarantee that the buffer remains valid for the duration of the application (e.g. string literal).
>  - The encoding is limited to Latin-1 instead of UTF-8.
>  - You must not call a destructor for the StringName.
> 
>  `is_static` is purely an optimization and can easily introduce undefined behavior if used wrong. In case of doubt, set it to false.
 

Since the semantics vary wildly (encoding, lifetime, destruction), would it make sense to split this into two functions?

* `string_name_new_with_utf8_chars`
* `string_name_new_static_with_latin1_chars`

(the "with_xy_chars" naming follows the `string_new_*` convention)

AThousandShips commented:
Note that a `StringName` constructed from a utf8 string that contains non-ascii characters will not match equality with one created from utf32, unless the utf8 one converts the buffer to utf32 before creating

Bromeon commented:
@dsnopek 

> > If a static StringName is created, its destructor should not be invoked?
> 
> I'm really not sure about this one. I think it may still be OK to invoke the destructor? I'm actually quite confused about how `StringName::_Data::static_count` is being used. This needs more investigation or an expert opinion.

Regarding that, last time I ran into problems when destroying static `StringName`s, and IIRC there weren't any memory leaks reported by LSan. Would need to check again, was more than a month ago...

---

@AThousandShips 

> Note that a `StringName` constructed from a utf8 string that contains non-ascii characters will not match equality with one created from utf32, unless the utf8 one converts the buffer to utf32 before creating

From UTF-32, so you mean converted from a `String`? Because the `char*` source here is either UTF-8 or Latin-1 (or even ASCII).

But that's an interesting limitation? Why is the buffer not normalized to one format?
Or are you talking about the static case (but then it would be Latin-1/ASCII)?

AThousandShips commented:
Meant `char*` with utf8, if it is utf8 it would have to be converted to utf32 to be matching with the hash, or done for the hash calculation

Bromeon commented:
But doesn't Godot use that internally with `SNAME` all the type?

Or that just happens to work because `SNAME` is used everywhere and not actually converted from `String`? But that is also hard to imagine, as user provided input like class names comes from `String`...

AThousandShips commented:
If you use `SNAME` it is not guaranteed to work if it contains non-ASCII

AThousandShips commented:
Reason being:
https://github.com/godotengine/godot/blob/75f9c97deaa8378477441e07ea1782ecab6d2eca/core/string/ustring.cpp#L2765-L2775C2

AThousandShips commented:
Causing for example "£" to be hashed as if it was "0xC2 0xA3" instead of "0xA3"

Bromeon commented:
> Note that a `StringName` constructed from a utf8 string that contains non-ascii characters will not match equality with one created from utf32, unless the utf8 one converts the buffer to utf32 before creating

So getting back to this, is it even a good idea to expose the UTF-8 based construction to GDExtension? You can be sure that users will run into this trap sooner or later...

Should we maybe just start by exposing the static ASCII constructor?
(I'm btw still not sure if ASCII or Latin-1...)

Or allow UTF-8 construction and convert to UTF-32 in the presence of non-ASCII chars? Might still be fast enough for the average case where `StringName` is pure ASCII.

AThousandShips commented:
I'd say to err on the side of caution and do conversion, arguably unconditionally, unless the performance benefits of having the shorter encoding is worth it

Bromeon commented:
Going always through `StringName` (3) would however defeat the point of this addition, as that's already possible with the existing API. The main idea behind this PR is to allow faster `StringName` construction, especially from literals.

Option (1) could still provide an `is_static` flag and could look as follows:
```cpp
string_name_new_with_latin1_chars(dest, c_string, is_static);
```

It would then be the caller's responsibility to make sure that `c_string` is Latin-1. 
This would be symmetric with the existing constructor for `String`, namely `string_new_with_latin1_chars`.

Later (or now?), we could then provide a constructor from UTF-8 that would unconditionally convert, [just like the `String` constructor](https://github.com/godotengine/godot/blob/75f9c97deaa8378477441e07ea1782ecab6d2eca/core/extension/gdextension_interface.cpp#L575).

AThousandShips commented:
Sounds good!

dsnopek commented:
> So either:
> 
> 1. Clarify that it only supports Latin-1 (encoded as such)
> 2. Check for encoding and convert conditionally
> 3. Convert unconditionally (i.e. just use a `String` conversion in the creation step):
>    `memnew_placement(r_dest, StringName(String::utf8(p_contents), static_cast<bool>(is_static)));`

I'd personally vote for option nr 1

`StringName`s are meant to be identifiers, not proper strings. So, in my opinion, sticking with ASCII is fine, it just happens that Godot treats C-strings as Latin-1 (as far as I can tell, at least) which is a mostly harmless superset of ASCII.

> Later (or now?), we could then provide a constructor from UTF-8 that would unconditionally convert, [just like the `String` constructor](https://github.com/godotengine/godot/blob/75f9c97deaa8378477441e07ea1782ecab6d2eca/core/extension/gdextension_interface.cpp#L575).

I'd vote for later. Given that it's already possible to convert a `String` to `StringName`, I don't think we're blocking anyone. Let's wait to see if anyone actually asks for this.

Bromeon commented:
> So, in my opinion, sticking with ASCII is fine, it just happens that Godot treats C-strings as Latin-1 (as far as I can tell, at least) which is a mostly harmless superset of ASCII.

I think this is where things break: while `StringName`s constructed from ASCII `char*` compare fine with the `StringName`s constructed through `String`, this doesn't hold for Latin-1 `char*`. See also [AThousandShips's comment here](https://github.com/godotengine/godot/pull/78580#issuecomment-1657186470) and the 2 messages preceding it.

Maybe the hash functions should be aligned? If yes, that would become blocking.

Or we only support ASCII and validate it. But in these constructor functions, there is currently no way to report errors. We could add a `GDExtensionBool` return type, but it would be inconsistent with other constructors like `string_new_*`.


---


> I'd vote for later. Given that it's already possible to convert a `String` to `StringName`, I don't think we're blocking anyone. We could wait to see if anyone actually asks for this.
> 
> (**UPDATE**: Although, if you really want to address it now, I'd be fine with that too :-))

I already implemented it in the draft 🙂 still needs some testing though.

dsnopek commented:
> I think this is where things break: while `StringName`s constructed from ASCII `char*` compare fine with the `StringName`s constructed through `String`, this doesn't hold for Latin-1 `char*`. See also [AThousandShips' comment here](https://github.com/godotengine/godot/pull/78580#issuecomment-1657186470) and the 2 messages preceding it.

Hm, looking at the code, I guess I still don't see why the hash is different for Latin-1 vs UTF-32.

The example of "£" is 0xA3 in Latin-1 and 0x000000A3 in UTF-32. I had thought this worked for all the Latin-1 characters? But I'm not an expert on this, so I could be wrong.

Looking at the hash function (for Latin-1 data):

```
uint32_t String::hash(const char *p_cstr) { 
 	uint32_t hashv = 5381; 
 	uint32_t c = *p_cstr++; 
  
 	while (c) { 
 		hashv = ((hashv << 5) + hashv) + c; /* hash * 33 + c */ 
 		c = *p_cstr++; 
 	} 
  
 	return hashv; 
 } 
```

It's casting each character to a `uint32_t` before operating on it, which I think effectively converts the Latin-1 character to UTF-32 (assuming I'm correct above, about Latin-1 being the same as the first 255 unicode characters).

Of course, if you passed UTF-8 data into `String::hash(const char *p_cstr)` then you'd get a very different hash, but I think this is meant to take Latin-1 data.

> Or we only support ASCII and validate it. But in these constructor functions, there is currently no way to report errors. We could add a `GDExtensionBool` return type, but it would be inconsistent with other constructors like `string_new_*`.

However, if we operate under the assumption that I'm wrong about the hashing above - which very well may be the case :-) - and we can only correctly handle ASCII, I think it'd be fine documenting that it only takes ASCII and naming it accordingly (ie. `string_name_new_with_ascii_chars()`), I don't think we necessarily need to validate it.

The binding will have more knowledge about the data coming from the developer, and may already have validated that the data is ASCII (perhaps the language even only supports ASCII?) and adding extra validation in the GDExtension interface is unnecessary. There's already definitely parts of the GDExtension interface that expect you to pass correct data in with no validation, so this wouldn't be anything new.

AThousandShips commented:
As I was saying, Latin-1, as in the encoding, should work correctly, but I don't know what encodings are used in source as C++ largely uses utf8 AFAIK, so that's my only point of concern, but assuming it is encoded as Latin-1 it should be fully safe

dsnopek commented:
Ah, ok, thanks for clarifying!

The encoding used by the end-developer is going to depend on their specific programming language, so, in my opinion, it should be up to the language binding to ensure that it's passing data with the right encoding to GDExtension.

dsnopek commented:
Discussed at the GDExtension meeting, and we think this API looks great! The only remaining question is if there is still an encoding issue. If not, this is probably good to move forward. If so, limiting the API to UTF-8 and ASCII (not Latin-1) would also be fine.

Bromeon commented:
Found the reason of the discrepancy between the hashes:

1. `StringName` constructed from `char*` (Latin-1) 
    * uses the `uint32_t String::hash(const char *p_cstr)` method
    * based on `char`, which has range `[-128, 127]` on Windows
    * `'±'` has value `-79`

2. `StringName` constructed from existing `String` (UTF-8)
   * uses the `uint32_t String::hash() const` method
   * based on `char32_t`
   * `U'±'` has value `177`

---

How should we proceed?
* Make sure the hashes are computed the same for all strings.
   * Would need to be careful that the one which is used for method registrations doesn't change, otherwise all method hashes are invalidated.
   * Any other breaking potential?
* Make sure that `StringName` at least uses a cosistent hash function for itself.

Bromeon committed:
Fix platform-dependent hashing for char* and wchar_t*

Since char/wchar_t can be either signed or unsigned, its conversion to uint32_t leads to different values depending on platform.
In particular, the same string represented as char* (Latin-1; StringName direct construction) or uint32_t (UTF-8; constructed
via String) previously resulted in different hashes.

Bromeon commented:
Thanks for the review, will change parameter names.

Added a 2nd commit that would modify the `char*` hash and makes the two compare equal in my Rust tests, but we need to see if it doesn't break anything else.

We also need to see how we would do it for `wchar_t*`, as it essentially suffers from the same problem. It's a bit more difficult, because unlike `char`, it's size can vary, so I'm not sure if truncating to `uint16_t` would always give deterministic results.

Bromeon committed:
More precise documentation for GDExtension String constructors (bytes vs. character count)

Bromeon commented:
Added a UTF-8 constructor that takes length, in addition to the one relying on null-termination. Should greatly enhance interoperability with other string types. (I can also remove the one expecting null-terminated `char*`, in case that's preferred).

Furthermore, I noticed the docs are imprecise -- the `p_size` argument sometimes means "number of bytes" (Latin-1, UTF-8) and sometimes "number of characters" (Wide, UTF-16, UTF-32). I renamed the latters to `p_char_count` and updated parameter docs accordingly. Does that make sense?

dsnopek commented:
> Found the reason of the discrepancy between the hashes

Aha, great catch! I wouldn't have thought to consider the signedness of character type.

> Would need to be careful that the one which is used for method registrations doesn't change, otherwise all method hashes are invalidated.

Yes, this is important, I don't want to have to map more hashes :-)

> Furthermore, I noticed the docs are imprecise -- the `p_size` argument sometimes means "number of bytes" (Latin-1, UTF-8) and sometimes "number of characters" (Wide, UTF-16, UTF-32). I renamed the latters to `p_char_count` and updated parameter docs accordingly. Does that make sense?

Sounds good to me!

> What's left is mostly the question **whether the hash for `wchar_t` should also be fixed**, and how (as it might be 2 or 4 bytes wide).

Could we do something with `constexpr`? For example:

```
    uint32_t val;
    if constexpr (sizeof(wchar_t) == 2) {
        val = static_cast<uint16_t>(input_char);
    } else {
        val = static_cast<uint32_t>(input_char);
    }
```

Bromeon commented:
> Could we do something with `constexpr`? 

Great idea! I actually ended up using `std::conditional_t` since the conversion was used in several places, and this adds much less code.

> Yes, this is important, I don't want to have to map more hashes :-)

CI would report incompatibility if that happened, correct?

dsnopek commented:
> Great idea! I actually ended up using `std::conditional_t` since the conversion was used in several places, and this adds much less code.

This code looks great to me!

However, we'll have to see how the core and production teams feel about using `std::conditional_t`, since it isn't used anywhere in Godot yet, and we try to limit how much of the standard library and newer C++ features we use. (We do already use `constexpr` in a few places, so that could be a fallback, if necessary - but your version is much shorter and easier to follow, so I'd personally prefer yours.)

> CI would report incompatibility if that happened, correct?

Yes, it would, and the tests have passed, so I think we're good :-)

Bromeon commented:
> However, we'll have to see how the core and production teams feel about using `std::conditional_t`, since it isn't used anywhere in Godot yet, and we try to limit how much of the standard library and newer C++ features we use.

I see! I changed it to `std::conditional<...>::type`, which is already used in several places and should thus be less controversial.

akien-mga commented:
Thanks!

beicause commented:
I think this PR added unnecessary functions for StringName constructing.

>So far, an indirection via String was necessary, causing at least 2 allocations and copies (String; String inside StringName). Since StringNames often refer to string literals, this allows them to be directly constructed from C strings.

This is not true. The internal data of StringName is still utf32 String, and String is COW, so String->StringName is cheap.
`gdextension_string_name_new_with_latin1_chars` and `gdextension_string_name_new_with_utf8_chars_and_len` are still c string -> String -> StringName, which has no performance benifit.

>It also provides the is_static flag: when the source has static storage duration, no copy/allocation will be needed. However, the extension developer needs to uphold this lifetime guarantee.

The `p_static` usage is misunderstood, see also https://github.com/godot-rust/gdext/issues/1306. I'm not sure if the use of `p_static` should be encouraged.

beicause commented:
Benchmark in godot-rust still shows a little performance improvement 🤔:
```rust
#[bench(repeat = 10000)]
fn string_name_utf8_gstring() -> StringName {
    let str = "a string";
    StringName::from(GString::from(str))
}

#[bench(repeat = 10000)]
fn string_name_utf8() -> StringName {
    use godot::sys;
    let str = "a string";
    let utf8 = str.as_bytes();

    // SAFETY: Rust guarantees validity and range of string.
    unsafe {
        StringName::new_with_string_uninit(|ptr| {
            sys::interface_fn!(string_name_new_with_utf8_chars_and_len)(
                ptr,
                utf8.as_ptr() as *const std::ffi::c_char,
                utf8.len() as i64,
            );
        })
    }
}

#[bench(repeat = 10000)]
fn string_name_cstr_gstring() -> StringName {
    use godot::sys;
    let c_str = c"a string";
    StringName::from(&GString::try_from_cstr(c_str, godot::builtin::Encoding::Latin1).unwrap())
}

#[bench(repeat = 10000)]
fn string_name_cstr() -> StringName {
    use godot::sys;
    let c_str = c"a string";
    unsafe {
        StringName::new_with_string_uninit(|ptr| {
            sys::interface_fn!(string_name_new_with_latin1_chars)(
                ptr,
                c_str.as_ptr(),
                sys::conv::SYS_FALSE, // p_is_static
            )
        })
    }
}
```

```
                                             min       median
   -- string_name_utf8_gstring   ...      0.075μs      0.078μs
   -- string_name_utf8           ...      0.061μs      0.063μs
   -- string_name_cstr_gstring   ...      0.076μs      0.079μs
   -- string_name_cstr           ...      0.058μs      0.060μs
```
